### DISCUSSION TITLE
[Question] Why setters from properties in attributes are allowed?

### DISCUSSION THREAD

**Discussion by kyoyama-kazusa** [dotnet/csharplang#3931] [General] [answered]
Suppose that we have an attribute named `CustomAttribute`, where a property `Property` contains both get method and set method, like this:
```csharp
public sealed class CustomAttribute : Attribute
{
    public int Property { get; set; }
}
```

In this case, the property `Property` can be modified in the runtime. However, if a member is marked `CustomAttribute`, they will be metadata in this assembly.
```csharp
[Custom(Property = 20)]
public sealed class Test
{
    // ...
}
```

So... When I use reflection to modify the value `20` (like below), it always fails to re-assign a new value. I wonder how I can do to modify that value, and why the setter can be allowed in attributes?

> I use reflection to modify the value like:
> ```csharp
> var inst = typeof(Test).GetCustomAttribute<CustomAttribute>()!;
> inst.Property = 60; // `inst` can be modified the value, but the next time we get the value, it will turn to 20 again.
> ```

Thank you.

**Comment by HaloFour** [comment, marked as answer]
The setter needs to exist for the runtime to set the property.  That setter only changes that instance, it doesn't rewrite the metadata in the assembly, which is why the value resets when you get the attribute through reflection.  If you don't want to have mutable properties in your attributes you can use a constructor instead.

**Reply by kyoyama-kazusa** [reply]
That sounds reasonable! Do you mean the setter is a bit like the C# 9 `init` setter, providing a way to set the value only in initialization? So if I want to modify the value **anytime**, I should change another way to implement on this point?

Now I know the setter guarantees the property is optional to assign, and they cannot be modified after initialized.

Thank you for your reply! 😃
